src/analysis/convex/cone/proper.lean

```diff
@@ -33,21 +39,132 @@ open continuous_linear_map filter
+@[simp, norm_cast] lemma coe_closure (K : convex_cone 𝕜 E) : (K.closure : set E) = closure K := rfl
+
+@[simp] protected lemma mem_closure {K : convex_cone 𝕜 E} {a : E} :
+  a ∈ K.closure ↔ a ∈ closure (K : set E) := iff.rfl
 
-protected lemma mem_closure {K : convex_cone ℝ E} {a : E} :
-  a ∈ K.closure ↔ a ∈ closure (K : set E) :=
-iff.rfl
+lemma closure_eq_iff_is_closed {K : convex_cone 𝕜 E} : K.closure = K ↔ is_closed (K : set E) :=
+⟨ (λ h, by rw [← closure_eq_iff_is_closed, ← coe_closure, h]),
+  (λ h, set_like.coe_injective $ closure_eq_iff_is_closed.2 h) ⟩
 
 end convex_cone
+
+/-- A proper cone is a convex cone `K` that is nonempty and closed. Proper cones have the nice
+property that the dual of the dual of a proper cone is itself. This makes them useful for defining
+cone programs and proving duality theorems. -/
+structure proper_cone (𝕜 : Type*) (E : Type*)
+  [ordered_semiring 𝕜] [add_comm_monoid E] [topological_space E] [has_smul 𝕜 E]
+  extends convex_cone 𝕜 E :=
+(nonempty'  : (carrier : set E).nonempty)
+(is_closed' : is_closed (carrier : set E))
+
+namespace proper_cone
+
+section has_smul
+
+variables {𝕜 : Type*} [ordered_semiring 𝕜]
+variables {E : Type*} [add_comm_monoid E] [topological_space E] [has_smul 𝕜 E]
+
+instance : has_coe (proper_cone 𝕜 E) (convex_cone 𝕜 E) := ⟨λ K, K.1⟩
+
+@[simp] lemma to_convex_cone_eq_coe (K : proper_cone 𝕜 E) : K.to_convex_cone = K := rfl
+
+lemma ext' : function.injective (coe : proper_cone 𝕜 E → convex_cone 𝕜 E) :=
+λ S T h, by cases S; cases T; congr'
+
+-- TODO: add convex_cone_class that extends set_like and replace the below instance
+instance : set_like (proper_cone 𝕜 E) E :=
+{ coe := λ K, K.carrier,
+  coe_injective' := λ _ _ h, proper_cone.ext' (set_like.coe_injective h) }
+
+@[ext] lemma ext {S T : proper_cone 𝕜 E} (h : ∀ x, x ∈ S ↔ x ∈ T) : S = T := set_like.ext h
+
+@[simp] lemma mem_coe {x : E} {K : proper_cone 𝕜 E} : x ∈ (K : convex_cone 𝕜 E) ↔ x ∈ K := iff.rfl
+
+protected lemma nonempty (K : proper_cone 𝕜 E) : (K : set E).nonempty := K.nonempty'
+
+protected lemma is_closed (K : proper_cone 𝕜 E) : is_closed (K : set E) := K.is_closed'
+
+end has_smul
+
+section module
+
+variables {𝕜 : Type*} [ordered_semiring 𝕜]
+variables {E : Type*} [add_comm_monoid E] [topological_space E] [t1_space E] [module 𝕜 E]
+
+instance : has_zero (proper_cone 𝕜 E) :=
+⟨ { to_convex_cone := 0,
+    nonempty' := ⟨0, rfl⟩,
+    is_closed' := is_closed_singleton } ⟩
+
+instance : inhabited (proper_cone 𝕜 E) := ⟨0⟩
+
+end module
+
+section inner_product_space
+
+variables {E : Type*} [normed_add_comm_group E] [inner_product_space ℝ E]
+variables {F : Type*} [normed_add_comm_group F] [inner_product_space ℝ F]
+
+protected lemma pointed (K : proper_cone ℝ E) : (K : convex_cone ℝ E).pointed :=
+(K : convex_cone ℝ E).pointed_of_nonempty_of_is_closed K.nonempty K.is_closed
+
+-- TODO: Replace map with a bundled version: proper_cone ℝ E →L[ℝ] proper_cone ℝ F
```

Comment:
This doesn't really make any sense right now; `proper_cone` is not currently a module and not currently a topological space. Are both these results true?

Reply:
Lean never gave me a "missing instance" error so I assumed that these instance might be present somewhere in the hierarchy.

Reply:
Similar definitions are there in the [dual cone section](https://leanprover-community.github.io/mathlib_docs/analysis/convex/cone/basic.html#the-dual-cone) for convex cones.

Reply:
Sorry, I misunderstood your question. Let me think about this ...

Reply:
Yes! Both of these statements are true (at least in the `inner_product_space` case). I'll update the TODO to reflect this.

Reply:
Possibly the `linear_map` version holds in greater generality.

Reply:
Yeah, that would even be true for `convex_cone`! Continuity is really only needed for the closedness requirement.

Reply:
Actually, never mind. I was being dumb. A convex cone is not a module (but a submodule is a convex cone). I'll get rid of these TODOs.